### PR TITLE
Fixed creating clusters in different namespaces when pivoting

### LINKED ISSUE #735: Creating clusters in non-default namespace using clusterctl fails at the pivot
/kind bug

**What steps did you take and what happened:**
1. Create machines.yaml and cluster.yaml.  Add a namespace attribute and give it a new scope.  Use clusterctl in a self-service workflow (using bootstrap and pivot).

**What did you expect to happen:**
New cluster would get created on the target cloud infrastructure and have machines and cluster objects created on the target cluster and scoped to the new namespace.

**Anything else you would like to add:**
Last two lines of the run,
```
I0205 16:36:38.368342   87810 kind.go:55] Ran: kind [delete cluster --name=clusterapi] Output:
F0205 16:36:38.368386   87810 create_cluster.go:59] unable to apply cluster api stack to target cluster: unable to pivot cluster API objects: error moving Cluster "test1": error creating cluster in namespace test: namespaces "test" not found
```

**Environment:**
vSphere but should also fail in other providers, based on the code in clusterdeployer.go.

- Cluster-api version:
- Kubernetes version: (use `kubectl version`): 1.11.3
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by sflxn**
There was a bug that prevented pivoting because a desired namespace was specified
in the cluster.yaml file.  It needs to ensure the the namespace on the target
cluster before pivoting.  Also, ensuring namespace after getting the target cluster
kubeconfig requires a loop with exp backoff in some environments.

Fixes #735

### CODE REVIEW

File: cmd/clusterctl/clusterdeployer/clusterdeployer.go [hunk 1]
```diff
@@ -102,6 +103,20 @@ func (d *ClusterDeployer) Create(cluster *clusterv1.Cluster, machines []*cluster
...
+		if err != nil {
+			return false, nil
+		}
+		return true, nil
+	}
+
+	err = util.Retry(addNamespaceToTarget, 0)
+	if err != nil {
+		return errors.Wrapf(err, "unable to ensure namespace %q in targetCluster", cluster.Namespace)
+	}
```
[the remark this answers is not in the document]
Comment (sflxn) [reply cmd/clusterctl/clusterdeployer/clusterdeployer.go]: That is my usual coding style, but the existing code used the one I have in my PR.  I find the practice of getting results from a method and checking the results in the same if statement a hotly debated topic.  I personally like the way that you stated, but I also try to use whatever prevailing coding style is in the existing code. :)

File: cmd/clusterctl/clusterdeployer/clusterdeployer.go [hunk 2]
```diff
@@ -102,6 +103,20 @@ func (d *ClusterDeployer) Create(cluster *clusterv1.Cluster, machines []*cluster
 	}
 	defer closeClient(targetClient, "target")
 
+	klog.Info("Creating namespace on target cluster")
```
[the remark this answers is not in the document]
Comment (detiber) [reply cmd/clusterctl/clusterdeployer/clusterdeployer.go]: +1 to including the namespace in the output

File: cmd/clusterctl/clusterdeployer/clusterdeployer.go [hunk 3]
```diff
@@ -102,6 +103,20 @@ func (d *ClusterDeployer) Create(cluster *clusterv1.Cluster, machines []*cluster
...
+			return false, nil
+		}
+		return true, nil
+	}
+
+	err = util.Retry(addNamespaceToTarget, 0)
+	if err != nil {
+		return errors.Wrapf(err, "unable to ensure namespace %q in targetCluster", cluster.Namespace)
+	}
+
```
[the remark this answers is not in the document]
Comment (detiber) [reply cmd/clusterctl/clusterdeployer/clusterdeployer.go]: +1 to moving after applying addonComponents

[the remark this answers is not in the document]
Comment (sflxn) [reply cmd/clusterctl/clusterdeployer/clusterdeployer.go]: It don't think it matters, but I think that doing it after we get the target client and before we do anything else to the target cluster is a much safer thing to do.  It is a good demarcation line for any new code in the future.  Any new operations that affects the target cluster is safe to add after the ensure namespace operation.  Putting the namespace operation after addons may make implementors of new operations guess whether they should put it before addons, after addons, or after namespace operation.

File: cmd/clusterctl/clusterdeployer/clusterdeployer.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (detiber) [reply cmd/clusterctl/clusterdeployer/clusterdeployer.go]: We've been moving towards standardizing on the more compact approach.

**Comment by detiber** [comment]
lgtm pending outstanding comments.

File: cmd/clusterctl/clusterdeployer/clusterdeployer.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sflxn) [reply cmd/clusterctl/clusterdeployer/clusterdeployer.go]: 2-to-1.  I'll move it.

**Pull request merged**